mne/decoding/ssd.py

```diff
@@ -290,3 +301,37 @@ def apply(self, X):
         pick_patterns = self.patterns_[self.sorter_spec][:self.n_components].T
         X = pick_patterns @ X_ssd
         return X
+
+
+def _dimensionality_reduction(cov_signal, cov_noise, info, rank):
+    """Perform dimensionality reduction on the covariance matrices."""
+    from scipy import linalg
+    n_channels = cov_signal.shape[0]
+    rank_signal = list(compute_rank(
+        Covariance(cov_signal, info.ch_names, list(), list(), 0,
+                   verbose=_verbose_safe_false()),
+        rank, _handle_default('scalings_cov_rank', None), info).values())[0]
+    rank_noise = list(compute_rank(
+        Covariance(cov_signal, info.ch_names, list(), list(), 0,
+                   verbose=_verbose_safe_false()),
+        rank, _handle_default('scalings_cov_rank', None), info).values())[0]
+    rank = np.min([rank_signal, rank_noise])  # should be identical
+
+    if rank < n_channels:
+        eigvals, eigvects = linalg.eigh(cov_signal)
+        # sort in descending order
+        ix = np.argsort(eigvals)[::-1]
+        eigvals = eigvals[ix]
+        eigvects = eigvects[:, ix]
+        # compute dimensionality reduction transformation matrix
+        dim_red = np.matmul(
+            eigvects[:, :rank], np.eye(rank) * (eigvals[:rank]**-0.5))
+        logger.info(
+            "Reducing covariance rank from %i -> %i" % (n_channels, rank,))
+    else:
+        dim_red = np.eye(n_channels)
+        logger.info("Preserving covariance rank (%i)" % (rank,))
+
+    cov_signal_red = np.matmul(dim_red.T, np.matmul(cov_signal, dim_red))
+    cov_noise_red = np.matmul(dim_red.T, np.matmul(cov_noise, dim_red))
+    return cov_signal_red, cov_noise_red, dim_red
```

Comment:
consider changing the name of the projection matrix from dim_red to proj_rank ? or something more accurate.

Reply:
Sure, it is a bit obtuse: perhaps I could change this as well.